### PR TITLE
Cleanup/shield refactor

### PR DESCRIPTION
**Pull request by z64555**
Combines shield management code from AI and Player sources into objectshield.cpp, which should provide consistent behavior between the two. At the very least, it reduces the number of places to maintain.

The balancing routine was generalized so that both AI and Players can use them, in their own form. Shield Equalization is just a balance with a rate of 100%.

The quadrant augment/transfer routine is a lightly modified flavor of the one zookeeper recently made for his model-point shields, namely to allow different rates to be plugged in.

### CODE REVIEW

**Comment by z64555** [comment]
A concern I'd like to raise regard model point shields, I honestly don't know if the shield_balance() function will behave with them.

**Comment by Goober5000** [comment]
I gave this a once-over, but it would be good to go through and compare the new functions side-by-side with the old ones using something like WinMerge, which I don't have time to do at the moment.

@z64555, can you be more specific with your concern about model point shields?  (Also, do we know who implemented that feature?)

**Comment by niffiwan** [comment]
@ln-zookeeper implemented model point shields

**Comment by z64555** [comment]
@Goober5000 I have concerns about it since I'm not sure how they interact with the existing shield system, that bit I haven't dug through yet. Also, I don't have a test model with model point shields to work with at the moment.

**Comment by z64555** [comment]
Ok, going to update this one with latest master. Still need a model with model-point shields to test behavior, though.

**Comment by z64555** [comment]
Alright! Rebased on top of master, and confirmed operation with stock models and one with modelpoint shields.

File: code/ai/aicode.cpp [hunk 1]
```diff
@@ -1999,9 +1999,9 @@ void evaluate_object_as_nearest_objnum(eval_nearest_objnum *eno)
 				num_attacking = num_enemies_attacking(OBJ_INDEX(eno->trial_objp));
                 
                 if (!sip->is_big_or_huge() && num_attacking < eno->max_attackers) {
-                    dist *= (float)(num_attacking + 2) / 2.0f;				//	prevents lots of ships from attacking same target
-                }
-				
+						dist *= (float) (num_attacking+2)/2.0f;				//	prevents lots of ships from attacking same target
```
Comment (The-E) [inline code/ai/aicode.cpp]: This whitespace change seems a bit weird

File: code/object/objectshield.cpp [hunk 2]
```diff
@@ -17,34 +17,69 @@
 
 #include <limits.h>
 
+// Private variables
+static double factor = 1.0 / (log(50.0) - log(1.0));	// Factor used in Goober5000's scale_quad
```
Comment (The-E) [inline code/object/objectshield.cpp]: I think "factor" is a bit too generic a name, especially if it is used as a global variable in this code. Renaming it to "shield_scale_factor" or something like that would make the intent clearer.

Comment (asarium) [inline code/object/objectshield.cpp]: Also, this variable seems to be a constant so it should be declared `const` to make sure no one changes it.

File: code/ai/aicode.cpp (hunk 1 above)
Comment (z64555) [inline code/ai/aicode.cpp]: Good catch, there seems to be numerous whitespace changes that has no business of being in this PR. I'll see if I can revert such changes.

**Comment by z64555** [comment]
Ok, needed to do a forced rebase to remove all those pesky whitespace edits that where not intended. No idea how they snuck in there, but I made sure that TortoiseDiff will display whitespace changes now.

Also pushed a commit that renamed `factor` to be `shield_scale_factor` and made it into a float, which might save a few clock cycles that would've been wasted on converting a double into a float.

**Comment by z64555** [comment]
@The-E 
> I do have a question though: Given that all the functions in here take a object pointer as their first parameter, wouldn't it make sense to turn them into member functions of the object class? That's a change that is probably out of scope for this PR, given that it would involve changes in a lot more places, but I would still like to see us make use of the fact that object is a class now.

That's a good point, and I agree that it should be a separate PR, this one's focused on merging the two shield management paths (one from AI, one from the player) into one path.

File: code/object/objectshield.cpp [hunk 3]
```diff
@@ -17,34 +17,69 @@
...
+ * @brief Logarithmically scales a quadrant's strength according to the shield generator's health %
+ *
+ * @param[in] generator_fraction The shield generator's HP in %
+ * @param[in] quad_strength      The quadrant's strength, before scaling
+ *
+ * @returns The quadrant's effective strength.
+ *
+ * @author Goober5000
+ */
+float scale_quad(float generator_fraction, float quad_strength);
```
Comment (asarium) [inline code/object/objectshield.cpp]: Why is this function declared here and the immediately defined a few lines after this? The declaration and definition could be merged here.

Comment (z64555) [inline code/object/objectshield.cpp]: It's a matter of style, I prefer to keep declarations of functions that are local to the .cpp separate from their definition.

Doing so makes it faster and easier for me to scroll through the definitions without being encumbered by the doxy comments. (I like to collapse function definitions in Visual Studio and in Notepadd++). I then keep the doxy comment as an "API" guide while implementation-specific comments are inside the definitions themselves.

**Comment by z64555** [comment]
`shield_transfer` Is called in two places, one by the AI code and the other by the player. Differences between the two (notably the different rates) are from retail behavior.

**Comment by asarium** [comment]
When I looked over the code the two places where you replaced the code used the code differently but it looks like your new `shield_transfer` function makes sure that the correct stuff is done depending on whether the object parameter refers to the Player object.

Alright, we now have two positive reviews so I guess this is ready to be merged.

**Pull request merged**